enterprise/server/remote_execution/uffd/uffd.go

```diff
@@ -248,27 +248,55 @@ func (h *Handler) handle(ctx context.Context, memoryStore *copy_on_write.COWStor
 			}
 			return status.InternalErrorf("read event from uffd failed with errno(%d)", err)
 		}
-		guestPageAddr := pageStartAddress(guestFaultingAddr, pageSize)
 
-		// Find the memory data in the store that should be used to handle the page fault
-		faultStoreOffset, err := guestMemoryAddrToStoreOffset(guestPageAddr, mappings)
+		mapping, err := guestMemoryAddrToMapping(uintptr(guestFaultingAddr), mappings)
 		if err != nil {
-			return status.WrapError(err, "translate to store offset")
+			return err
 		}
-		relOffset := memoryStore.GetRelativeOffsetFromChunkStart(faultStoreOffset)
+
+		guestPageAddr := pageStartAddress(guestFaultingAddr, pageSize)
+		if guestPageAddr < mapping.BaseHostVirtAddr {
+			// Make sure we only try to map addresses that fall within the valid
+			// guest memory ranges
+			guestPageAddr = mapping.BaseHostVirtAddr
+		}
+
+		// Find the memory data in the store that should be used to handle the page fault
+		faultStoreOffset := guestMemoryAddrToStoreOffset(guestPageAddr, *mapping)
+
+		// To reduce the number of UFFD round trips, try to copy the entire
+		// chunk containing the faulting address
 		hostAddr, allocatedLen, err := memoryStore.GetChunkStartAddressAndSize(uintptr(faultStoreOffset), false /*=write*/)
 		if err != nil {
 			return status.WrapError(err, "get backing page address")
 		}
 
+		// If copying the entire chunk would map data falling outside the valid
+		// guest memory range, only copy the valid parts of the chunk
+		relOffset := memoryStore.GetRelativeOffsetFromChunkStart(faultStoreOffset)
+		destAddr := guestPageAddr - relOffset
+		// Check for data below the valid memory range
+		if destAddr < mapping.BaseHostVirtAddr {
+			invalidBytesAtChunkStart := mapping.BaseHostVirtAddr - destAddr
+			destAddr = mapping.BaseHostVirtAddr
+			hostAddr += invalidBytesAtChunkStart
+			allocatedLen -= int64(invalidBytesAtChunkStart)
+		}
+		// Check for data above the valid memory range
+		lastValidAddress := mapping.BaseHostVirtAddr + mapping.Size - 1
+		copyEndAddress := destAddr + uintptr(allocatedLen)
+		if copyEndAddress > lastValidAddress {
+			invalidBytesAtChunkEnd := int64(copyEndAddress - lastValidAddress)
+			allocatedLen -= invalidBytesAtChunkEnd
+		}
```

Comment:
it looks like there might be an off-by-one error in computing `invalidBytesAtChunkEnd` here because `lastValidAddress` is an inclusive end index (start + length - 1) but `mappingEndAddr` is an exclusive end index (start + length), and when those two things are subtracted I think the result is off by one:

```suggestion
		mappingEndAddr := mapping.BaseHostVirtAddr + mapping.Size
		copyEndAddr := destAddr + uintptr(allocatedLen)
		if copyEndAddr > mappingEndAddr {
			invalidBytesAtChunkEnd := int64(copyEndAddr - mappingEndAddr)
			allocatedLen -= invalidBytesAtChunkEnd
		}
```

Reply:
Ah thanks :) Saved me from having to find this